File: 2nd_quarter/week04/rash/신고 결과 받기.py
def solution(id_list, report, k):
    # 계획
    # 신고한 사람을 저장하는 딕셔너리 생성
    # 신고당한 횟수를 저장하는 딕셔너리 생성
    # 중복 신고 제외
    # 신고 횟수가 k를 넘으면 answer리스트 +1
    answer = [0] * len(id_list)
    arrcnt = {}
    repcnt = {}
    for i in range(len(id_list)):
        arrcnt[id_list[i]] = 0
        repcnt[id_list[i]] = []

    processed_reports = set()

    for i in report:
        tmp = i.split(" ")
        reporter = tmp[0]
        reported = tmp[1]

        if i in processed_reports:
            continue

        processed_reports.add(i)

        if reported in arrcnt:
            arrcnt[reported] += 1
            repcnt[reporter].append(reported)

    for i in range(len(id_list)):
        reported_users = repcnt[id_list[i]]
        for user in reported_users:
            if arrcnt[user] >= k:
                answer[i] += 1

    return answer


## 개쩐다고 생각한 풀이
def solution(id_list, report, k):
    answer = [0] * len(id_list)
    reports = {}
    for x in id_list:
        reports[x] = 0

    for r in set(report):
        reports[r.split()[1]] += 1

    for r in set(report):
        if reports[r.split()[1]] >= k:
            answer[id_list.index(r.split()[0])] += 1

    return answer

File: 2nd_quarter/week04/Royce/신고 결과 받기.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <set>

using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    vector<int> answer(id_list.size(), 0);
    // 개수를 0으로 할당해준다.
    
    map<string, int> mp;
    map<string, set<string>> mp_report;
    // 신고한 아이디를 중복을 없애줘야하므로 set으로 처리한다.
    
    for(int i=0;i<id_list.size();i++){
        mp[id_list[i]] = i;
        // 각각의 스트링에 번호를 부여한다.
    }
    
    for(auto re : report){
        stringstream ss(re);
        string a, b;
        ss >> a >> b;
        
        mp_report[b].insert(a);
        // a가 b를 신고하였으므로, b에 a를 넣어줌
        // [신고당한 id, 신고한 id]
    }
    
    for(auto re : mp_report){
        if(re.second.size() >= k){   // k.second는 신고당한 사람에 관해 신고한 사람의 수
            for(auto ch : re.second){   // 즉 second일 경우 신고한 사람들이므로
                answer[mp[ch]]++;       // 그에 관하여 메일을 다 보내주었다.
            }
            // 위에서 mp[ch]에 관해서 번호를 할당해야 answer에서 그에 알맞은 인덱스에 추가됨
        }
    }
    
    
    return answer;
}

